astacus/coordinator/coordinator.py

```diff
@@ -221,31 +246,82 @@ def get_subresult_url(request_url: URL, op_id: int) -> str:
-        state: CoordinatorState = Depends(coordinator_state)
-    ):
-        self.request = request
-        self.request_url = request.url
-        self.background_tasks = background_tasks
-        self.config = config
-        self.state = state
-        self.stats = statsd.StatsClient(config=config.statsd)
-
-        assert self.config.object_storage
-        mstorage = MultiRohmuStorage(config=self.config.object_storage)
-        self.hexdigest_mstorage = mstorage
-        json_mstorage: MultiStorage = mstorage
-        if self.config.object_storage_cache:
-            file_mstorage = MultiFileStorage(self.config.object_storage_cache)
-            json_mstorage = MultiCachingJsonStorage(backend_mstorage=mstorage, cache_mstorage=file_mstorage)
-        self.json_mstorage = json_mstorage
-        self.sync_lock = utils.get_or_create_state(state=request.app.state, key="sync_lock", factory=threading.RLock)
+    def __init__(self, *, c: Coordinator, op_id: int, attempts: int, steps: List[Step]):
+        super().__init__(c=c, op_id=op_id)
+        self.state = c.state
+        self.attempts = attempts
+        self.steps = steps
+        self.step_progress = {}
+
+    @property
+    def progress(self) -> Progress:
+        return Progress.merge(self.step_progress.values())
+
+    async def run_with_lock(self, cluster: Cluster) -> None:
+        name = self.__class__.__name__
+        try:
+            for attempt in range(1, self.attempts + 1):
+                logger.debug("%s - attempt #%d/%d", name, attempt, self.attempts)
+                context = StepsContext(attempt=attempt)
+                async with self._timing_manager("astacus_attempt_duration", {"op": name, "attempt": str(attempt)}):
+                    try:
+                        if await self.try_run(cluster, context):
+                            return
+                    except exceptions.TransientException as ex:
+                        logger.info("%s - transient failure: %r", name, ex)
+        except exceptions.PermanentException as ex:
+            logger.info("%s - permanent failure: %r", name, ex)
+        self.set_status_fail()
+
+    async def try_run(self, cluster: Cluster, context: StepsContext) -> bool:
+        op_name = self.__class__.__name__
+        for i, step in enumerate(self.steps, 1):
+            step_name = step.__class__.__name__
+            if self.state.shutting_down:
+                logger.info("Step %s not even started due to shutdown", step_name)
+                return False
+            logger.debug("Step %d/%d: %s", i, len(self.steps), step_name)
+            async with self._timing_manager("astacus_step_duration", {"op": op_name, "step": step_name}):
+                with self._progress_handler(cluster, step):
+                    r = await step.run_step(cluster, context)
+            if not r:
+                logger.info("Step %s failed", step)
+                return False
+            context.set_result(step.__class__, r)
```

Comment:
I think you should also ensure that they key is already not set, just in case.

Reply:
Fixed in the StepsContext.